File: log-simulator/port_scan.py
import random
import time

def generate_port_scan_logs():
    attacker_ip = f"172.16.50.{random.randint(100, 199)}"
    target_ip = "192.168.1.10"
    common_ports = [21, 22, 23, 25, 53, 80, 110, 139, 443, 445, 1433, 3306, 3389, 8080]
    logs = []
    
    # Simulate a port scan on 5-8 random common ports
    scanned_ports = random.sample(common_ports, k=random.randint(5, 8))
    for port in scanned_ports:
        timestamp = time.strftime("%Y-%m-%d %H:%M:%S")
        action = "REJECT" if port not in [80, 443] else "ALLOW"
        logs.append(f"{timestamp} fw-gateway-1: src={attacker_ip} dst={target_ip} proto=TCP spt={random.randint(49152, 65535)} dpt={port} action={action} bytes=0")
        
    return logs

if __name__ == "__main__":
    for log in generate_port_scan_logs():
        print(log)

File: log-simulator/src/attacks/portScan.ts
import { SecurityLogEntry } from '../types/log';
import { generateExternalIp } from '../utils/ipGenerator';
import { randomSample, randomInt } from '../utils/random';
import { COMMON_PORTS } from '../utils/constants';

/**
 * Simulates a port scan attack sequence.
 * Returns an array of firewall logs representing rejected connections on multiple ports.
 */
export function simulatePortScan(): SecurityLogEntry[] {
  const logs: SecurityLogEntry[] = [];
  const attackerIp = generateExternalIp();
  const targetIp = '192.168.1.10'; // Internal DMZ server
  const baseTime = new Date();
  
  // Select 6-10 ports to scan
  const portsToScan = randomSample(COMMON_PORTS, randomInt(6, 10)).map(p => p.port);

  portsToScan.forEach((port, idx) => {
    // Spacer: nmap scans ports very fast (milliseconds apart)
    const timestamp = new Date(baseTime.getTime() + idx * randomInt(50, 200));
    
    // Format: "YYYY-MM-DD HH:mm:ss"
    const formattedDate = timestamp.toISOString().replace('T', ' ').substring(0, 19);
    
    // Most common port scan behavior blocks/rejects traffic
    const action = 'REJECT'; 
    const spt = randomInt(49152, 65535);

    const msg = `${formattedDate} fw-gateway-1: src=${attackerIp} dst=${targetIp} proto=TCP spt=${spt} dpt=${port} action=${action} bytes=0`;

    logs.push({
      timestamp,
      source: 'firewall',
      logLevel: 'WARNING',
      message: msg,
      details: {
        src_ip: attackerIp,
        dst_ip: targetIp,
        proto: 'TCP',
        port,
        dst_port: port,
        action,
        bytes: 0
      }
    });
  });

  return logs;
}
